File: pive/visualization/chordchart.py
# Copyright (c) 2014 - 2015, David Bothe
# All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:

# 1. Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.

# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import jinja2
import json
from pive.visualization import defaults as default
from pive.visualization import basevisualization as bv

from pathlib import Path


class Chart(bv.BaseVisualization):
    def __init__(self,
                 dataset,
                 template_name,
                 width=default.width,
                 height=default.height,
                 padding=default.padding):

        # Initializing the inherited pseudo-interfaces.
        bv.BaseVisualization.__init__(self)

        # Metadata
        self._title = 'chordchart'
        self._template_name = 'chordchart'
        self._dataset = dataset
        self._dataset_url = ''

        self._template_url = Path(__file__).resolve().parent.joinpath(default.template_path)
        self._datakeys = []
        self._version = default.p_version
        self._js_name = "Chordchart"

        # Visualization properties.
        self._width = width
        self._height = height
        self._padding = padding
        self._colors = default.chartcolors
        self._label_size = default.label_size

        #Axis properties.
        self._shape_rendering = default.shape_rendering
        self._line_stroke = default.line_stroke
        self._font_size = default.font_size

        #Chord specific.
        self._textpadding = default.textpadding
        self._elementfontsizse = default.fontsize
        self._tickfontsize = default.ticksize
        self._ticksteps = default.ticksteps
        self._tickprefix = default.prefix

    def set_title(self, title):
        self._title = title

    def setDataKeys(self, datakeys):
        self._datakeys = datakeys;

    def setTicksteps(self, ticksteps):
        self._ticksteps = ticksteps;

    def setTickprefix(self, tickprefix):
        self._tickprefix = tickprefix;

    def set_chart_colors(self, colors):
        """Basic Method."""
        self._colors = colors

    def get_modifiable_template_variables(self):
        """Returns a dictionary of all template variables, that are supposed to be modifiable by the client.
        Subclasses should override this method and add their own variables.
        """

        variables = super().get_modifiable_template_variables()
        variables["t_datakeys"] = self._datakeys
        variables["t_textpadding"] = self._textpadding
        variables["t_elementfontsizse"] = self._elementfontsizse
        variables["t_tickfontsize"] = self._tickfontsize
        variables["t_ticksteps"] = self._ticksteps
        variables["t_tickprefix"] = self._tickprefix
        return variables

    def get_modifiable_template_variables_typehints(self):
        """Returns a dictionary of typehints for variables that are modifiable by the client.
        Subclasses should override this method and add their own variables.
        """
        typehints = super().get_modifiable_template_variables_typehints();
        new_typehints = {
            "default" : {
                "t_datakeys": {
                    "type": "list",
                    "length": len(self._datakeys),
                    "item_type": {
                        "type": "string"
                    }
                },
                "t_textpadding": {
                    "type": "int",
                    "min": 1
                },
                "t_elementfontsizse": {
                    "type": "int",
                    "min": 1
                },
                "t_tickfontsize": {
                    "type": "int",
                    "min": 1
                },
                "t_tickprefix": {
                    "type": "string",
                    "min": 1,
                    "max": 1
                }
            }
        }
        for key in new_typehints.keys():
            if key not in typehints.keys():
                typehints[key] = {}
            typehints[key].update(new_typehints[key])
        return typehints

    def initMatrixRow(self, elements):
        """Initializes a matrix row with zero values."""
        mrow = []
        for i in range(0, elements):
            mrow.append(0)
        return mrow

    def generateAdjacencyMatrix(self, elements, dataset):
        """Generates the adacency matrix of the graph."""
        matrix = []
        for source in elements:
            i = 0

            mrow = self.initMatrixRow(len(elements))

            for line in dataset:

                keys = list(line.keys())

                current_element = line[keys[0]]
                if source == current_element:

                    for dest in elements:
                        if line[keys[1]] == dest:
                            index = elements.index(dest)
                            mrow.pop(index)
                            mrow.insert(index, line[keys[2]])
                            i += 1
            matrix.append(mrow)
        return matrix

    def generate_visualization_dataset(self, dataset):
        """Basic Method."""
        visdataset = {}

        elements = []

        for item in dataset:
            keys = list(item.keys())
            # The elements consist of the distinct sources
            # and destinations.
            source = item[keys[0]]
            dest = item[keys[1]]
            if source not in elements:
                elements.append(source)
            if dest not in elements:
                elements.append(dest)
        matrix = self.generateAdjacencyMatrix(elements, dataset)

        visdataset['elements'] = elements
        visdataset['matrix'] = matrix

        return visdataset


    def create_js(self, template, dataset_url):
        templateVars = {'t_width': self._width,
                        't_height': self._height,
                        't_padding': self._padding,
                        't_url': dataset_url,
                        't_colors': self._colors,
                        't_textpadding': self._textpadding,
                        't_elementFontSize': self._elementfontsizse,
                        't_tickFontSize': self._tickfontsize,
                        't_ticksteps': self._ticksteps,
                        't_tickprefix': self._tickprefix,
                        't_div_hook': self._div_hook,
                        't_font_size': self._font_size,
                        't_shape_rendering': self._shape_rendering,
                        't_line_stroke': self._line_stroke,
                        't_pive_version' : self._version,
                        't_axis_label_size' : self._label_size}

        outputText = template.render(templateVars)
        return outputText

    def load_from_dict(self, dictionary):
        super().load_from_dict(dictionary)
        if "t_datakeys" in dictionary:
            self.setDataKeys(json.loads(dictionary['t_datakeys'].replace('\'', '\"')))
        if "t_textpadding" in dictionary:
            self._textpadding = int(dictionary['t_textpadding'])
        if "t_elementfontsizse" in dictionary:
            self._elementfontsizse = dictionary['t_elementfontsizse']
        if "t_tickfontsize" in dictionary:
            self._tickfontsize = dictionary['t_tickfontsize']
        if "t_ticksteps" in dictionary:
            self.setTicksteps( int(dictionary['t_ticksteps']) )
        if "t_tickprefix" in dictionary:
            self.setTickprefix((dictionary['t_tickprefix']))

File: pive/visualization/static/chordchart.js
/* Copyright (c) 2014 - 2015, David Bothe
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice,
 this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.

*/

class Chordchart {
	constructor(config, init) {
		this.config = config;
		this.serialisable_elements = [
			'width', 'height', 'padding', 'textpadding', 'elementfontsize', 'tickfontsize', 'ticksteps', 'tickprefix', 'div_hook',
			'timeformat', 'iconwidth', 'iconheight', 'iconcolor', 'iconhighlight', 'iso', 'scales', 'colors',
			'label_size'
		];
		this.width = config.width;
		this.height = config.height;
		this.padding	= config.padding;
		this.textpadding = config.textpadding;
		this.elementfontsize = config.elementfontsize;
		this.tickfontsize = config.tickfontsize;
		this.ticksteps = config.ticksteps;
		this.tickprefix = config.tickprefix;
		this.url = config.dataset_url;
		this.div_hook = config.div_hook;
		this.colors = config.colors;
		this.label_size = config.label_size;

		this.innerRadius = Math.min(this.width, this.height) * .32;
		this.outerRadius = this.innerRadius * 1.2;
		this.labelFont = "Helvetica";
		this.transistionSpeed = 500;

		this.init = init;
	}

	get_current_config() {
		let config = {};
		for (let element_name of this.serialisable_elements) {
			config[element_name] = this[element_name];
		}
		return config;
	}


	render() {
		var root_div = document.getElementById(this.div_hook);
		root_div.innerHTML = "";
		
		const css_line = `#${this.div_hook} .line { stroke: ${this.config.line_stroke}; fill: none; stroke-width: 2.5px}\n`,
			css_tooltip = `#${this.div_hook} .tooltip {color: white; line-height: 1; padding: 12px; font-weight: italic; font-family: arial; border-radius: 5px;}\n`,
			css_axis_path = `#${this.div_hook} .axis path { fill: none; stroke: ${this.config.line_stroke}; shape-rendering: crispEdges;}\n`,
			css_axis_line = `#${this.div_hook} .axis line { stroke: ${this.config.line_stroke}; shape-rendering: ${this.config.shape_rendering };}\n`,
			css_path_area = `#${this.div_hook} .path area { fill: blue; }\n`,
			css_axis_text = `#${this.div_hook} .axis text {font-family: sans-serif; font-size: ${this.config.font_size }px }\n`,
			css_xlabel_text = `#${this.div_hook} .xlabel {font-family: helvetica; font-size: ${this.label_size }px }\n`,
			css_ylabel_text = `#${this.div_hook} .ylabel {font-family: helvetica; font-size: ${this.label_size }px }\n`,
			css_x_axis_line = `#${this.div_hook} .x.axis line { stroke: grey; stroke-opacity: 0.25; stroke-width: 2.5px}\n`,
			css_y_axis_line = `#${this.div_hook} .y.axis line { stroke: grey; stroke-opacity: 0.25; stroke-width: 2.5px}`;

		var css = css_line.concat(css_tooltip).concat(css_axis_path).concat(css_axis_line).concat(css_path_area).concat(css_axis_text)
				  .concat(css_xlabel_text).concat(css_ylabel_text).concat(css_x_axis_line).concat(css_y_axis_line);

		var style = document.createElement('style');
		style.type = 'text/css';
		style.appendChild(document.createTextNode(css));
		root_div.appendChild(style);
		const chart_object = this;
		d3.json(chart_object.url, chart_object.init).then(function (data) {
			var svg = d3.select('#'.concat(chart_object.div_hook)).append("svg")
			.attr("width", chart_object.width)
			.attr("height", chart_object.height)
			.append("g")
			.attr("transform", "translate(" + chart_object.width/2 + "," + chart_object.height/2 + ")");

			var chordElements = data.elements;
			var matrix = data.matrix;

			function getColorIndex(index) {
				var colorindex = index;
				var color;

				while (colorindex > chart_object.colors.length - 1) {
					colorindex = colorindex - chart_object.colors.length;

				}

				color = chart_object.colors[colorindex];
				return color;
			}

			function generateChord() {

				svg.selectAll("rect").remove();
				svg.selectAll("g").remove();

				svg.attr("transform", "translate(" + chart_object.width / 2 + "," + chart_object.height / 2 + ")");

				var chord = d3.chord()
					//.matrix(matrix)
					//.padding(0.05)
					.sortSubgroups(d3.descending);

				//Select all groups from class "chordGroup"
				var chordGroup = svg.selectAll("g.chordGroup")
					.data(chord(matrix).groups)
					.enter().append("svg:g")
					.attr("class", "chordGroup");

				var arc = d3.arc()
					.innerRadius(chart_object.innerRadius)
					.outerRadius(chart_object.outerRadius);

				var ribbon = d3.ribbon()
					.radius(chart_object.innerRadius)

				chordGroup.append("path")
					.attr("d", arc)
					.style("fill", function (d) {
						return getColorIndex(d.index);
					})
					.style("stroke", function (d) {
						return getColorIndex(d.index);
					})
					.attr("id", function (d, i) {
						return "group-" + d.index;
					});

				chordGroup.append("text")
					.each(function (d) {
						d.angle = (d.startAngle + d.endAngle) / 2;
					})
					.attr("dy", ".35em")
					.attr("font-size", chart_object.elementfontsize)
					.attr("font-family", chart_object.labelFont)
					.attr("transform", function (d) {
						return "rotate(" + (d.angle * 180 / Math.PI - 90) + ")"
							+ "translate(" + (chart_object.outerRadius + chart_object.textpadding) + ")"
							+ (d.angle > Math.PI ? "rotate(180)" : "");
					})
					.style("text-anchor", function (d) {
						return d.angle > Math.PI ? "end" : null;
					})
					.text(function (d) {
						return chordElements[d.index];
					});


				//Eventlistener for each chordGroup group.
				chordGroup.on("mouseover", fade(0.05))
					.on("mouseout", fade(0.75))
					.on("click", function (d, i) {
						showDetail(i.index);
					});


				svg.append("g")
					.attr("class", "chord")
					.selectAll("path")
					.data(chord(matrix))
					.enter().append("path")
					.attr("d", ribbon)
					.style("fill", function (d) {
						return chordColor(d);
					})
					.style("stroke", function (d) {
						return chordColor(d);
					})
					.style("opacity", 0.75);

				var ticks = svg.append("g").selectAll("g")
					.data(chord(matrix).groups)
					.enter().append("g").selectAll("g")
					.data(chordTicks)
					.enter().append("g")
					.attr("class", "ticks")
					.attr("transform", function (d) {
						return "rotate(" + (d.angle * 180 / Math.PI - 90) + ")"
							+ "translate(" + chart_object.outerRadius + ",0)";
					});


				ticks.append("line")
					.attr("x1", 1)
					.attr("y1", 0)
					.attr("x2", 5)
					.attr("y2", 0)
					.style("stroke", "#111");

				ticks.append("text")
					.attr("x", 8)
					.attr("dy", ".35em")
					.attr("font-size", chart_object.tickfontsize)
					.attr("font-family", chart_object.labelFont)
					.attr("transform", function (d) {
						return d.angle > Math.PI ? "rotate(180)translate(-16)" : null;
					})
					.style("text-anchor", function (d) {
						return d.angle > Math.PI ? "end" : null;
					})
					.text(function (d) {
						return d.label;
					});

			}

			function chordColor(d) {

				var colorindex = (d.source.value > d.target.value ?
					d.source.index : d.target.index);
				return getColorIndex(colorindex);
			}


			function fade(opacity) {
				return function (d, i) {
					svg.selectAll(".chord path")
						.filter(function (d) {
							return d.source.index != i.index &&
								d.target.index != i.index;
						})
						.transition("fade")
						.duration(chart_object.transistionSpeed)
						.style("opacity", opacity);
				};
			};

			//Draws a detailed BarChart containing the absolute weight of the graphnode.
			function showDetail(index) {

				svg.attr("transform", "translate(0,0)");

				svg.selectAll("g").remove()

				const detailwidth = chart_object.width / 1.5;
				const detailheight = chart_object.height / 2;

				var color = getColorIndex(index);

				var dataset = matrix[index];

				var xScale = d3.scaleBand()
					.domain(d3.range(dataset.length))
					.rangeRound([0, detailwidth - chart_object.padding])
					.padding(0.1);

				var yScale = d3.scaleLinear()
					.domain([0, d3.max(dataset)])
					.range([0, detailheight - chart_object.padding]);

				var xAxisScale = d3.scaleBand()
					.domain(chordElements)
					.rangeRound([0, detailwidth - chart_object.padding])
					.padding(0.1);

				var yAxisScale = d3.scaleLinear()
					.domain([0, d3.max(dataset)])
					.range([detailheight - chart_object.padding, 0]);


				//Define X axis
				var xAxis = d3.axisBottom()
					.scale(xAxisScale)
					.ticks(5);

				//Define Y axis
				var yAxis = d3.axisLeft()
					.scale(yAxisScale)
					.ticks(10);

				//Create X axis
				svg.append("g")
					.attr("opacity", 0.0)
					.attr("class", "x axis")
					// .attr("transform", "translate(0," + (detailheight - padding) + ")")
					.attr("transform", "translate(" + chart_object.padding + "," + (detailheight - chart_object.padding) + ")")
					.call(xAxis)
					.transition("createXAxis")
					.duration(chart_object.transistionSpeed)
					.attr("opacity", 1.0);

				//Create Y axis
				svg.append("g")
					.attr("opacity", 0.0)
					.attr("class", "y axis")
					.attr("transform", "translate(" + chart_object.padding + ",0)")
					.call(yAxis)
					.transition("createYAxis")
					.duration(chart_object.transistionSpeed)
					.attr("opacity", 1.0);

				var bars = svg.selectAll("rect")
					.data(dataset)
					.enter()
					.append("rect")
					.attr("opacity", 0.0)
					.attr("fill", color)
					.on("click", generateChord)
					.transition("drawBars")
					.duration(chart_object.transistionSpeed)
					.attr("class", "bar")
					.attr("opacity", 1.0)
					.attr("x", function (d, i) {
						return xScale(i) + chart_object.padding;
					})
					.attr("y", function (d) {
						return detailheight - yScale(d) - chart_object.padding;
					})
					.attr("width", xScale.bandwidth())
					.attr("height", function (d) {
						return yScale(d);
					});
			};


			// Returns tick angles.
			function chordTicks(d) {
				var k = (d.endAngle - d.startAngle) / d.value;

				return d3.range(0, d.value, chart_object.ticksteps).map(function (v, i) {
					return {
						angle: v * k + d.startAngle,
						label: i % 5 ? null : v / chart_object.ticksteps + chart_object.tickprefix
					};
				});
			};
			generateChord();
		});

	}
}

pive.addChartVersion("chordchart", "0.3.4", Chordchart);